File: solutions/0846. Hand of Straights/0846.py
# Time complexity: O(sort)
# Space complexity: O(n)
class Solution:
  def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
    if len(hand) % groupSize != 0:
      return False

    cardCount = Counter(hand)
    for card in sorted(hand):
      if cardCount[card] == 0:
        continue

      for i in range(groupSize):
        if cardCount[card + i] > 0:
          cardCount[card + i] -= 1
        else:
          return False

    return True

File: solutions/0846. Hand of Straights/0846.cpp
// Time complexity: O(sort)
// Space complexity: O(n)
class Solution {
 public:
  bool isNStraightHand(vector<int>& hand, int groupSize) {
    if (hand.size() % groupSize != 0)
      return false;

    ranges::sort(hand);

    unordered_map<int, int> cardCount;
    for (const int card : hand)
      ++cardCount[card];

    for (const int card : hand) {
      if (cardCount[card] == 0)
        continue;

      for (int i = 0; i < groupSize; ++i)
        if(cardCount[card + i] > 0)
          --cardCount[card + i];
        else
          return false;
    }

    return true;
  }
};